openverse-api/catalog/api/views/media_views.py

```diff
@@ -0,0 +1,190 @@
+    Swagger documentation.
+
+    :param field_names: the list of field names to convert to Markdown
+    :return: the names as a Markdown string
+    """
+
+    *all_but_last, last = field_names
+    all_but_last = ', '.join([f'`{name}`' for name in all_but_last])
+    return f'{all_but_last} and `{last}`'
+
+
+def _get_user_ip(request):
+    """
+    Read request headers to find the correct IP address.
+    It is assumed that X-Forwarded-For has been sanitized by the load balancer
+    and thus cannot be rewritten by malicious users.
+    :param request: A Django request object.
+    :return: An IP address.
+    """
+    x_forwarded_for = request.META.get('HTTP_X_FORWARDED_FOR')
+    if x_forwarded_for:
+        ip = x_forwarded_for.split(',')[0]
+    else:
+        ip = request.META.get('REMOTE_ADDR')
+    return ip
+
+
+class SearchMedia(APIView):
+    swagger_schema = CustomAutoSchema
+    search_description = (
+        """
+Although there may be millions of relevant records, only the most
+relevant several thousand records can be viewed. This is by design:
+the search endpoint should be used to find the top 10,000 most relevant
+results, not for exhaustive search or bulk download of every barely
+relevant result. As such, the caller should not try to access pages
+beyond `page_count`, or else the server will reject the query.
+
+For more precise results, you can go to the
+[Openverse Syntax Guide](https://search.creativecommons.org/search-help)
+for information about creating queries and
+[Apache Lucene Syntax Guide](https://lucene.apache.org/core/2_9_4/queryparsersyntax.html)
+for information on structuring advanced searches.
+"""  # noqa
+        f'{refer_sample}'
+    )
+
+    def _get(self,
+             request,
+             default_index, qa_index,
+             query_serializer, media_serializer, result_serializer):
+        params = query_serializer(data=request.query_params)
+        if not params.is_valid():
+            return input_error_response(params.errors)
+
+        hashed_ip = hash(_get_user_ip(request))
+        page_param = params.data[PAGE]
+        page_size = params.data[PAGESIZE]
+        qa = params.data[QA]
+        filter_dead = params.data[FILTER_DEAD]
+
+        search_index = qa_index if qa else default_index
+        try:
+            results, num_pages, num_results = search_controller.search(
+                params,
+                search_index,
+                page_size,
+                hashed_ip,
+                request,
+                filter_dead,
+                page=page_param
+            )
+        except ValueError as value_error:
+            return input_error_response(value_error)
+
+        context = {'request': request}
+        serialized_results = media_serializer(
+            results,
+            many=True,
+            context=context
+        ).data
+
+        if len(results) < page_size and num_pages == 0:
```

Comment:
I'm not sure it belongs to this PR or a separate one:
The `num_pages` returns a number that is 1 less than the actual number of pages. For example, if the `page_size` is 20, and we only have 40 results, `num_pages` will be 1, although in reality we get 2 pages. If there are fewer than 20 results, there will be 0 pages?

Reply:
The search controller calculates page size as follows:
https://github.com/WordPress/openverse-api/blob/eefd9a46d09f97ba8c7077d5734772a5611cf50e/openverse-api/catalog/api/controllers/search_controller.py#L487

This calculation always leads to page count less by 1. We can fix this but there will break backwards compatibility with API clients that add 1 to the page count from their end.

Reply:
We're not going to be transferring API users from CC to Openverse, just notifying them that there's a new API available, so I think _now_ is a reasonable time to make this fix. We can note the behavior in our email to current API users.

Reply:
Cool, I'll make this a separate issue as many PRs are chaining up behind this one.